src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -33,25 +37,44 @@ export class CellExecutionQueue {
     }
     constructor(
         private readonly notebookPromise: Promise<INotebook>,
-        private readonly executionFactory: CellExecutionFactory,
+        private readonly cellExecutionFactory: CellExecutionFactory,
+        private readonly codeExecutionFactory: CodeExecutionFactory,
         readonly metadata: Readonly<KernelConnectionMetadata>
     ) {}
     /**
      * Queue the cell for execution & start processing it immediately.
      */
-    public queueCell(cell: NotebookCell): void {
-        const existingCellExecution = this.queueOfCellsToExecute.find((item) => item.cell === cell);
-        if (existingCellExecution) {
-            traceCellMessage(cell, 'Use existing cell execution');
-            return;
-        }
-        const cellExecution = this.executionFactory.create(cell, this.metadata);
-        this.queueOfCellsToExecute.push(cellExecution);
+    public queueCell(code: NotebookCell | string): Promise<nbformat.IOutput[]> {
```

Comment:
I couldn't figure out how to return both types (`Promise<NotebookCellRunState[] | nbformat.IOutput[]>`) here and then use `instanceof nbformat.IOutput[]` on `kernelExecution.ts`.

If you do, let me know @DonJayamanne

Reply:
I'd created two methods

Reply:
Yeah, and as is the naming is quite awkward here at least in my view. It's called queueCell, but it takes a parameter called 'code' now not cell. And then the paths below are harder to read and easy to miss whens on. @DonJayamanne 's suggestion is good. If it was one method it would be better with a more explicit name like queueExecution, then casting early to either CodeExecution or CellExecution so that the code after that is easier to read, but two methods sounds like a legit idea.

Reply:
+1 on a separate method